### PR TITLE
Testsuite for user projects

### PR DESCRIPTION
**Pull request by tamiko**
This is a major refactoring of the testsuite code to allow user projects to
easily define tests as is done in deal.II. A typical example is a user
project building a library "bob" with tests located in a subdirectory tests:

  ./tests/foo.cc
  ./tests/foo.output

And a

  ./tests/CMakeLists.txt

consisting of

  SET(TEST_LIBRARIES bob)
  DEAL_II_PICKUP_TESTS()

This will set up (depending on build type Debug, Release, DebugRelease) the
tests foo.debug and/or foo.release. Behaving similarly as tests in our
testsuite.

The documentation update will be done in a separate pull request.

b596ce0 (Matthias Maier, 61 minutes ago)
   Minor cleanup

1ca1834 (Matthias Maier, 32 minutes ago)
   CMake: Bugfix: Force cached CMAKE_BUILD_TYPE in testsuite

5692dd8 (Matthias Maier, 2 hours ago)
   CMake: Improve DEAL_II_ADD_TEST macro

    - For every deal.II build type (given by the variable DEAL_II_BUILD_TYPES)
     that is a (case insensitive) substring of CMAKE_BUILD_TYPE a test is
     defined.

83208c9 (Matthias Maier, 3 hours ago)
   CMake: Add "DEBUG|RELEASE" parameter to DEAL_II_SETUP_TARGET

   It is now possible to force compilation and linkage against a specific
   flavor of deal.II by using "DEBUG" or "RELEASE" as keywords, e.g.,

     DEAL_II_SETUP_TARGET(target DEBUG)

01ebd33 (Matthias Maier, 4 hours ago)
   CMake: Refactor testsuite setup into DEAL_II_PICKUP_TESTS

   This commit refactors all the testsuite setup that was originally done in
   setup_testsuitesubproject.cmake into the DEAL_II_PICKUP_TESTS macro defined
   in ./cmake/macros/macro_deal_ii_pickup_tests.cmake

8eb17b3 (Matthias Maier, 4 hours ago)
   CMake: Relocate testsuite macros and scripts to ./cmake/macros

   This commit relocates all testsuite macros back to cmake/macros and the
   run_test.cmake and normalize.pl script to cmake/scripts. Further, all of
   those are copied to the build directory and installed appropriately.

07ccc69 (Matthias Maier, 5 hours ago)
   Change macro reldir to "share/deal.II/cmake"

### CODE REVIEW

**Comment by luca-heltai** [comment]
This looks very nice. In particular, with this changes, the project https://github.com/luca-heltai/bare-dealii-app is no longer needed, since one can do everything from withing the standard deal.II installation.

**Comment by tamiko** [comment]
Please wait with the merge. I have to resolve one last bug.

**Comment by luca-heltai** [comment]
Matthias, is there a way to ensure that you can run ctest in the `build` directory, instead of having to go to `build/tests`?

**Comment by tamiko** [comment]
If you call <code>ENABLE_TESTING()</code> in the top-level <code>CMakeLists.txt</code> file you can run ctest also in the build directory.

File: tests/setup_testsubproject.cmake [hunk 1]
```diff
@@ -1,140 +1,17 @@
...
-##
-## The deal.II library is free software; you can use it, redistribute
-## it, and/or modify it under the terms of the GNU Lesser General
-## Public License as published by the Free Software Foundation; either
-## version 2.1 of the License, or (at your option) any later version.
-## The full text of the license can be found in the file LICENSE at
-## the top level of the deal.II distribution.
-##
-## ---------------------------------------------------------------------
+FIND_PACKAGE(deal.II 8.0 REQUIRED HINTS ${DEAL_II_DIR})
```
Comment (bangerth) [inline tests/setup_testsubproject.cmake]: Did you remove the copyright notice on purpose?

**Comment by bangerth** [comment]
By the way, where is this documented? I've tried to look here
  http://dealii.org/developer/index.html
but there is no link, nor does it appear to be discussed in
  http://dealii.org/developer/users/cmakelists.html
or did I miss it somewhere?

File: tests/setup_testsubproject.cmake (hunk 1 above)
Comment (tamiko) [inline tests/setup_testsubproject.cmake]: Am 27. Mar 2015, 19:37 schrieb Wolfgang Bangerth <notifications@github.com>:

>> @@ -1,140 +1,17 @@
>> -## ---------------------------------------------------------------------
>> -##
>> -## Copyright (C) 2013 by the deal.II authors
>> -##
>> -## This file is part of the deal.II library.
>> -##
>> -## The deal.II library is free software; you can use it, redistribute
>> -## it, and/or modify it under the terms of the GNU Lesser General
>> -## Public License as published by the Free Software Foundation; either
>> -## version 2.1 of the License, or (at your option) any later version.
>> -## The full text of the license can be found in the file LICENSE at
>> -## the top level of the deal.II distribution.
>> -##
>> -## ---------------------------------------------------------------------
>> +FIND_PACKAGE(deal.II 8.0 REQUIRED HINTS ${DEAL_II_DIR})
>
> Did you remove the copyright notice on purpose?

No, this is a minor refactoring accident.

**Comment by tamiko** [comment]
> By the way, where is this documented? I've tried to look here
>   http://dealii.org/developer/index.html
> but there is no link, nor does it appear to be discussed in
>   http://dealii.org/developer/users/cmakelists.html
> or did I miss it somewhere?

As originally stated in the pull request, there will be a major
documentation in the next days. :-]

For now the quick guide is to put some test.(cc|output) into a user
project and call DEAL_II_PICKUP_TESTS() in this very directory.

Best,
Matthias

**Comment by bangerth** [comment]
OK, thanks. Just wondering because one of my students happens to need this right now -- you have good timing :-)

**Comment by luca-heltai** [comment]
@bangerth , I just updated  https://github.com/luca-heltai/bare-dealii-app to use the new mechanism. Take a look at the CMakeLists.txt there, to see how this

**Pull request merged**